crates/red_knot_python_semantic/src/semantic_index.rs

```diff
@@ -71,20 +72,25 @@ pub fn public_symbol<'db>(
 
 /// The symbol tables for an entire file.
 #[derive(Debug)]
-pub struct SemanticIndex {
+pub struct SemanticIndex<'db> {
     /// List of all symbol tables in this file, indexed by scope.
-    symbol_tables: IndexVec<FileScopeId, Arc<SymbolTable>>,
+    symbol_tables: IndexVec<FileScopeId, Arc<SymbolTable<'db>>>,
 
     /// List of all scopes in this file.
     scopes: IndexVec<FileScopeId, Scope>,
 
     /// Maps expressions to their corresponding scope.
     /// We can't use [`ExpressionId`] here, because the challenge is how to get from
     /// an [`ast::Expr`] to an [`ExpressionId`] (which requires knowing the scope).
-    scopes_by_expression: FxHashMap<NodeKey, FileScopeId>,
+    scopes_by_expression: FxHashMap<ExpressionNodeKey, FileScopeId>,
 
-    /// Map from the definition that introduce a scope to the scope they define.
-    scopes_by_definition: FxHashMap<NodeWithScopeKey, FileScopeId>,
+    /// Maps from a node creating a definition node to its definition.
+    definitions_by_node: FxHashMap<DefinitionNodeKey, Definition<'db>>,
+
+    /// Map from nodes that create a scope to the scope they create.
+    scopes_by_node: FxHashMap<NodeWithScopeKey, FileScopeId>,
+
+    scope_ids_by_scope: IndexVec<FileScopeId, ScopeId<'db>>,
```

Comment:
This replaces the `scopes_map`. I don't think it gives us much when we have more fine-grained queries in general